runtime/resetting_histogram.go

```diff
@@ -0,0 +1,298 @@
+func (NilResettingHistogram) Update(int64) {}
+
+// Clear is a no-op.
+func (NilResettingHistogram) Clear() {}
+
+func (NilResettingHistogram) Count() int64 {
+	panic("Count called on a NilResettingHistogram")
+}
+
+func (NilResettingHistogram) Variance() float64 {
+	panic("Count called on a NilResettingHistogram")
+}
+
+func (NilResettingHistogram) Min() int64 {
+	panic("Min called on a NilResettingHistogram")
+}
+
+func (NilResettingHistogram) Max() int64 {
+	panic("Max called on a NilResettingHistogram")
+}
+
+func (NilResettingHistogram) Sum() int64 {
+	panic("Sum called on a NilResettingHistogram")
+}
+
+func (NilResettingHistogram) StdDev() float64 {
+	panic("StdDev called on a NilResettingHistogram")
+}
+
+func (NilResettingHistogram) Sample() Sample {
+	panic("Sample called on a NilResettingHistogram")
+}
+
+func (NilResettingHistogram) Percentiles([]float64) []float64 {
+	panic("Percentiles called on a NilResettingHistogram")
+}
+
+func (NilResettingHistogram) Percentile(float64) float64 {
+	panic("Percentiles called on a NilResettingHistogram")
+}
+
+func (NilResettingHistogram) Mean() float64 {
+	panic("Mean called on a NilResettingHistogram")
+}
+
+// StandardResettingHistogram is used for storing aggregated values for timers, which are reset on every flush interval.
+type StandardResettingHistogram struct {
+	values []int64
+	mutex  sync.Mutex
+}
+
+func (t *StandardResettingHistogram) Count() int64 {
+	panic("Count called on a StandardResettingHistogram")
+}
+
+func (t *StandardResettingHistogram) Max() int64 {
+	panic("Max called on a StandardResettingHistogram")
+}
+
+func (t *StandardResettingHistogram) Min() int64 {
+	panic("Min called on a StandardResettingHistogram")
+}
+
+func (t *StandardResettingHistogram) StdDev() float64 {
+	panic("StdDev called on a StandardResettingHistogram")
+}
+
+func (t *StandardResettingHistogram) Variance() float64 {
+	panic("Variance called on a StandardResettingHistogram")
+}
+
+func (t *StandardResettingHistogram) Sum() int64 {
+	panic("Sum called on a StandardResettingHistogram")
+}
+
+func (t *StandardResettingHistogram) Sample() Sample {
+	panic("Sample called on a StandardResettingHistogram")
+}
+
+func (t *StandardResettingHistogram) Percentiles([]float64) []float64 {
+	panic("Percentiles called on a StandardResettingHistogram")
+}
+
+func (t *StandardResettingHistogram) Percentile(float64) float64 {
+	panic("Percentile called on a StandardResettingHistogram")
+}
+
+func (t *StandardResettingHistogram) Mean() float64 {
+	panic("Mean called on a StandardResettingHistogram")
+}
+
+// Values returns a slice with all measurements.
+func (t *StandardResettingHistogram) Values() []int64 {
+	return t.values
+}
+
+// Snapshot resets the timer and returns a read-only copy of its sorted contents.
+func (t *StandardResettingHistogram) Snapshot() Histogram {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	currentValues := t.values
+	t.values = make([]int64, 0, InitialResettingHistogramSliceCap)
```

Comment:
`t.values` is transferred to the `Snapshot`, and we initialise a new slice for the `Histogram`. If you are suggesting to re-use the memory of `t.values` for the next values in the Histogram, we have to copy them over for the Snapshot, which is essentially the same thing.

Reply:
I should've clarified. We could set a finalizer **on the `Snapshot`** that returns the slice to the pool of this histogram. Basically I'm thinking that the pattern will be:

* Long-lived resetting histogram.
* Every 5 seconds, call `Snapshot()` => transfers `t.values` onto the Snapshot and makes a new slice.
* The `Snapshot` gets used, probably immediately and just once. If we add a `Close()` method that returns the slice (reset to 0 length) to the histogram, so that it's available to be re-used the next time we rotate, we will effectively used as many slices as Snapshots are alive/retained (probably just 1 at a time) + the active slice, max. 2 or 3 slices.

With the current code, we will just allocate and GC naïvely.

Maybe this is premature optimisation, but wanted to raise this as I think it might bite us in the future.

Reply:
I see your point now. I think allocating a bit of memory (specially when we don't have thousands of histograms in our plans) a few times every 5-10sec. is fine.

We can totally do a memory pool, but feels like premature optimisation.

I suspect that when we go to TaaS and we start keeping longer retention for metrics, we might downgrade the intervals to 10sec. or even 30sec. intervals, which should reduce memory allocations even further.

Reply:
We can leave it like this for now. I opened an issue to track: https://github.com/testground/testground/issues/1021.